keps/sig-auth/2784-csr-duration/README.md

```diff
@@ -0,0 +1,473 @@
+
+1. Alice creates a certificate signing request with `spec.expirationSeconds` set to `600`
+1. The request is approved
+1. The signer issues the signed certificate for `600` seconds
+1. Alice is able to use the certificate for `10` minutes, after which it expires
+
+#### Story 2
+
+1. A privileged component `X` issues certificates for clients using the CSR API
+1. Based on its internal policy, it creates CSRs with `spec.expirationSeconds` set to `900`
+1. `X` approves the CSR and fetches the certificate once the signer issues it
+1. `X` validates that the signer honored the `spec.expirationSeconds` field
+1. `X` gives the certificate to the client
+1. The certificate automatically expires in `15` minutes
+
+### Notes/Constraints/Caveats (Optional)
+
+N/A
+
+### Risks and Mitigations
+
+This functionality will serve to reduce risk and increase security in the
+Kubernetes ecosystem by helping client migrate away from long lived certificates.
+
+In the worst case scenario, the new field will be ignored, which does not reduce
+security from the status quo.
+
+## Design Details
+
+This design is centered around a change to the `CertificateSigningRequestSpec`
+structs found in the `k8s.io/api/certificates/v1` and `k8s.io/api/certificates/v1beta1`
+packages.
+
+A new optional `ExpirationSeconds` field will be added to this struct.  The go
+doc comment describes the behavior of this field.  A `*uint32` is used because
+the field is optional, must be positive and must not overflow JSON parsers.
+
+```go
+// CertificateSigningRequestSpec contains the certificate request.
+type CertificateSigningRequestSpec struct {
+  // ... other fields omitted for brevity
+
+  // go doc omitted for brevity
+  SignerName string `json:"signerName" protobuf:"bytes,7,opt,name=signerName"`
+
+  // expirationSeconds is the requested duration of validity of the issued
+  // certificate. The certificate signer may issue a certificate with a different
+  // validity duration so a client must check the delta between the notBefore and
+  // and notAfter fields in the issued certificate to determine the actual duration.
+  //
+  // The v1.22+ in-tree implementations of the well-known Kubernetes signers will
+  // honor this field as long as the requested duration is not greater than the
+  // maximum duration they will honor per the --cluster-signing-duration CLI
+  // flag to the Kubernetes controller manager.
+  //
+  // Certificate signers may not honor this field for various reasons:
+  //
+  //   1. Old signer that is unaware of the field (such as the in-tree
+  //      implementations prior to v1.22)
+  //   2. Signer whose configured maximum is shorter than the requested duration
+  //   3. Signer whose configured minimum is longer than the requested duration
+  //
+  // The minimum valid value for expirationSeconds is 600, i.e. 10 minutes.
+  //
+  // +optional
+  ExpirationSeconds *uint32 `json:"expirationSeconds,omitempty" protobuf:"varint,8,opt,name=expirationSeconds"`
```

Comment:
What error do you get when someone sends you -1?  Is it a prototbuf / json decoding error with a 429?  If not, I'd probably prefer having int32 so that we can properly validate.

Reply:
agree... I can't find any other examples of uint in our rest API

Reply:
Changed to `*int32`.